### PR TITLE
Refactor CI/CD

### PR DESCRIPTION
**Pull request by KartikSoneji**
Refactor the CI/CD workflow
- Cache dependences and build outputs
- Build binary only once
- Release artifacts for each commit and PR to allow for easy testing
- Build and test on all platforms

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -20,7 +20,13 @@ jobs:
 
     steps:
       - name: Checkout
-        uses: actions/checkout@v2
+        uses: actions/checkout@v3
+
+      - name: Install libudev-dev on Ubuntu
+        if: matrix.os == 'ubuntu-latest'
```
Comment (samuelvanderwaal) [inline .github/workflows/build.yml:26]: This needs to be `buildjet-8vcpu-ubuntu-2004` because I upgraded to using BuildJet for Ubuntu runners.

**Review by samuelvanderwaal — CHANGES_REQUESTED** [review]
How much does caching speed up the builds? The reason we don't currently build on every branch is the Windows and MacOs build take 15-20 minutes and that's really onerous to wait on when you have a simple PR fix you're trying to merge. 

Otherwise this looks good except for we need to either manually change the Linux binary name to `ubuntu-latest` or we need to update the install script to use the BuildJet runner name.

**Comment by KartikSoneji** [comment]
> How much does caching speed up the builds? The reason we don't currently build on every branch is the Windows and MacOs build take 15-20 minutes and that's really onerous to wait on when you have a simple PR fix you're trying to merge.

If the builds take too long, we can mark the non-main branch ones as an optional check.
But in theory, it should be as fast as an incremental compile.

> Otherwise this looks good except for we need to either manually change the Linux binary name to `ubuntu-latest` or we need to update the install script to use the BuildJet runner name.

Oh right, I'll rename the binary.

**Comment by samuelvanderwaal** [comment]
> > How much does caching speed up the builds? The reason we don't currently build on every branch is the Windows and MacOs build take 15-20 minutes and that's really onerous to wait on when you have a simple PR fix you're trying to merge.
> 
> If the builds take too long, we can mark the non-main branch ones as an optional check. But in theory, it should be as fast as an incremental compile.
> 
> > Otherwise this looks good except for we need to either manually change the Linux binary name to `ubuntu-latest` or we need to update the install script to use the BuildJet runner name.
> 
> Oh right, I'll rename the binary.

Ok, once the conflict is resolved I'll run the action a few times and see how long it takes.

File: .github/workflows/build.yml (hunk 1 above)
Comment (KartikSoneji) [reply .github/workflows/build.yml]: Done.

**Review by samuelvanderwaal — APPROVED** [review]

**Comment by samuelvanderwaal** [comment]
Any idea why CI isn't triggering for this?

**Comment by KartikSoneji** [comment]
> Any idea why CI isn't triggering for this?

I don't think it will unless it's a branch in the same repo.

**Comment by samuelvanderwaal** [comment]
> > Any idea why CI isn't triggering for this?
> 
> I don't think it will unless it's a branch in the same repo.

This [one ran](https://github.com/metaplex-foundation/sugar/actions/runs/3207290152).

**Pull request merged**